Sources/Logging/Logging.swift

```diff
@@ -498,11 +498,34 @@ public struct MultiplexLogHandler: LogHandler {
     }
 }
 
+/// A wrapper to facilitate `print`-ing to stderr and stdio
+private struct FileOutputStream: TextOutputStream {
+    let file: UnsafeMutablePointer<FILE>
+
+    func write(_ string: String) {
+        string.withCString { ptr in
+            flockfile(file)
+            defer {
+                funlockfile(file)
+            }
+            _ = fputs(ptr, file)
+        }
+    }
+}
+
 /// Ships with the logging module, really boring just prints something using the `print` function
-internal struct StdoutLogHandler: LogHandler {
+public struct StdioLogHandler: LogHandler {
+
+    public enum Stream {
```

Comment:
@allenhumphreys just as a question if you think this is useful. We could also do the following here: Instead of an `enum Stream` we could do

```swift
// this is just to prevent name clashes further down
#if os(macOS) || os(tvOS) || os(iOS) || os(watchOS)
let sysStderr = Darwin.stderr
let sysStdout = Darwin.stdout
#else
let sysStderr = Glibc.stderr
let sysStdout = Glibc.stdout
#endif

public struct Stream {
    private let underlying: UnsafePointer<FILE>

    internal init(underlying: UnsafePointer<FILE>) {
        self.underlying = underlying
    }

    public static let stderr: Stream = .init(underlying: sysStderr)
    public static let stdout: Stream = .init(underlying: sysStdout)
}
```

that way, the API is the same for the user: `StdioLogHandler(label: "foo", stream: .stdout)` will still work but we have two extra benefits:

1. in tests, we can use `StdioLogHandler(label: "foo", stream: .init(underlying: fopen(".....", "w+")))`
2. if we chose to allow other streams in the future, then we can add them _without breaking public API_. The problem with `enum`s is: Adding a case breaks the API because someone might have done an exhaustive `switch` over it. The `struct` + `static let` solution from above alleviates that issue.

Reply:
Thanks @weissi , this is good, but I'd like to propose going a little further. 
1. Make `Stream` wrap `TextOutputStream` instead
2. Rename `StdioLogHandler` to `StreamLogHandler`
3. Go ahead and expose the initializer for `Stream`

It'd look something like this:
```
public struct StreamLogHandler: LogHandler {

    public struct Stream: TextOutputStream {

        private let underlying: TextOutputStream

        public init(underlying: TextOutputStream) {
            self.underlying = underlying
        }

        public mutating func write(_ string: String) {
            var underlying = self.underlying
            underlying.write(string)
        }

        public static let stderr: Stream = .init(underlying: FileOutputStream(file: sysStderr))
        public static let stdout: Stream = .init(underlying: FileOutputStream(file: sysStdout))
    }

    private let lock = Lock()
    public let stream: Stream

    public init(label: String, stream: Stream = .stdout) {
        self.stream = stream
    }

...
...
}
```

`FileOutputStream` would remain internal, but clearly available as a reference for consumers. This would allow people to immediately swap in their own streams. It eliminates the need to work with Pointers, or even the c standard library at all. `Stream` ends up's up just being a box and a convenient way to get access to `stderr` and `stdout` in a way that hides the `UnsafeMutablePointer<FILE>` stuff.

Thoughts? Too far?

Reply:
@allenhumphreys Thanks, that sounds like a great idea to me. I think you can remove the `Lock` because if the user wants locking, they could implement that themselves, no?

Reply:
Seems like a good idea 👍

Reply:
> @allenhumphreys Thanks, that sounds like a great idea to me. I think you can remove the Lock because if the user wants locking, they could implement that themselves, no?

Is that a safe default? in case a concurrent program would use the default loggers shipped?

Reply:
@ktoso I'd leave it up to the `TextOutputStream` to lock. That's at least what default `print` does: It uses a `TextOutputStream` that uses `f(un)lockfile`.

Reply:
If we instead keep `init(underlying: TextOutputStream)` internal for this PR, could we punt on the locking question?

Reply:
I suppose if we did that, the `FileOutputStream` is still handling the locking, which means removing it at the log handler level should happen either way. We could have a `LockingTextOutputStream` to make sure consumers of the API know that they should provide locking to avoid interleaving. Or having something like `init(underlying: TextOutputStream & Lockable)`

Reply:
+1 on removing the lock and leaving it as a concern of the `TextOutputStream`. the one we ship, `FileOutputStream`, should obv do the right thing, but if users choose to write their own they should know that they need to deal with concurrency, and we should document this at the API + readme level

Reply:
@tomerd @weissi @ktoso Finally looked at making the change. The lock in question was never actually used to lock the output, it is used to lock access to the properties on the `StreamLogHandler` itself, e.g. `metadata` and `logLevel`. `StreamLogHandler.Stream` will not do any locking of its output, the underlying `TextOutputStream` can do that. I'll make callouts for API adopters that they should consider locking if they implement their own.

Reply:
@allenhumphreys ah, you can remove that lock. `LogHandler`s are supposed to have value semantics for the configuration (metadata & logLevel) and indeed your handler has, so there's no need to lock.

Reply:
@weissi Thanks for pointing that out, it looks like the lock was a holdover from when `StdoutLogHandler` was a class, changed at https://github.com/apple/swift-log/commit/1d983153b45a6b6c90a8ebfab58d4d3e23350949

Reply:
@allenhumphreys thank you! Apologies, for some reason I thought you added the lock. I'd propose to just remove it because it's baggage from the past :).

Reply:
In light of discovering how `Swift.print` functions, I tweaked the API again. I removed the wrapper type `StreamLogHandler.Stream` and exposed `StdioOutputStream` to allow easily switching between `stderr` and `stdio`. `StdioOutputStream` is not publicly instantiable, though, so the usage is still restricted. I believe this API is now much more straightforward as it just uses `TextOutputStream`!

Reply:
@allenhumphreys perfect. I think it you make the output streams non-public, then we can merge this.